<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>add any phylogenetically based transcriptome evolutionary index (TEI) to
    single-cell data objects • scTEI</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="add any phylogenetically based transcriptome evolutionary index (TEI) to
    single-cell data objects">
<meta property="og:description" content="scTEI add any phylogenetically based transcriptome evolutionary
    index (TEI) to single-cell data objects">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">scTEI</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/scTEI.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kullrich/scTEI/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="sctei-">scTEI <a href="https://github.com/kullrich/scTEI" class="external-link"><img src="reference/figures/logo.png" align="right" height="160"></a>
<a class="anchor" aria-label="anchor" href="#sctei-"></a>
</h1></div>
</div>
<div class="section level1">
<h1 id="sctei---add-any-phylogenetically-based-transcriptome-evolutionary-index-tei-to-single-cell-data-objects">scTEI - add any phylogenetically based transcriptome evolutionary index (TEI) to single-cell data objects<a class="anchor" aria-label="anchor" href="#sctei---add-any-phylogenetically-based-transcriptome-evolutionary-index-tei-to-single-cell-data-objects"></a>
</h1>
</div>
<div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h1>
<p>The goal of <code>scTEI</code> is to provide easy functionality to add phylogenetically based transcriptome evolutionary index (TEI) to single-cell data objects.</p>
<p>For a comprehensive overview about the topic of gene age assignments, transcriptome age index (TAI) and its derivates (TDI, TPI, Adjusted SD, PhastCons, …) see e.g.:</p>
<ul>
<li><a href="https://drostlab.github.io/myTAI/articles/Introduction.html" class="external-link">myTAI - Introduction</a></li>
<li>TAI: (<a href="https://doi.org/10.1016/j.tig.2007.08.014" class="external-link">Domazet-Lošo, Brajković, and Tautz 2007</a>)</li>
<li>TAI: (<a href="https://doi.org/10.1038/nature09632" class="external-link">Domazet-Lošo and Tautz 2010</a>)</li>
<li>TDI: (<a href="https://doi.org/10.1038/nature11394" class="external-link">Quint et al. 2012</a>)</li>
<li>TDI: (<a href="https://doi.org/10.1093/molbev/msv012" class="external-link">Drost et al. 2017</a>)</li>
<li>TPI: (<a href="https://doi.org/10.1093/molbev/msw044" class="external-link">Gossmann et al. 2016</a>)</li>
<li>Adjusted SD, PhastCons: (<a href="https://doi.org/10.1186/s12915-020-00842-z" class="external-link">Liu et al. 2020</a>)</li>
</ul>
<p><strong>Note that prior calculating TEI one needs to retrieve phylogenetic or taxonomic information for your focal species.</strong></p>
<p>This might be a <em>phylostratigraphic map</em> (as introduced by Domazet-Lošo, Brajković, and Tautz 2007) or an <em>ortho map</em> (I call them, see e.g. Julca et al. 2021 or Cazet et al. 2022), which can be obtained by assigning to each orthogroup (OG) or hierachical orthogroup (HOG) along a given species tree the ancestral node.</p>
<p>Please have a look at the introduction of the great myTAI package for possible sources, how to get such <em>phylostratigraphic maps</em> (<a href="https://drostlab.github.io/myTAI/articles/Introduction.html" class="external-link">myTAI - Introduction</a>).</p>
<p>To create an <em>ortho map</em>, one “simply” needs to get OGs or HOGs with e.g. OrthoFinder (Emms and Kelly 2019) or Proteinortho (Lechner et al. 2011) or any other ortholog prediction tool (see Linard et al. 2021) using a set of species that cover the species range of your interest. Parse each OG or HOG for the oldest clade as compared to a species tree and your focal species of interest.</p>
<p>Or e.g. use pre-calculated OGs from e.g. <a href="https://omabrowser.org/oma/home" class="external-link uri">https://omabrowser.org/oma/home</a> (Schneider, Dessimoz, and Gonnet 2007) or <a href="http://eggnog5.embl.de" class="external-link uri">http://eggnog5.embl.de</a> (Huerta-Cepas et al. 2019) or <a href="https://bioinformatics.psb.ugent.be/plaza/" class="external-link uri">https://bioinformatics.psb.ugent.be/plaza/</a> for plants (Proost et al. 2009).</p>
<p>Example of <em>Caenorhabditis elegans</em>:</p>
<p>single-cell data: Packer and Zhu et al. 2019</p>
<p>ortho map: Sun et al. 2021</p>
<p><img src="reference/figures/celegans_seurat_p2_p3.png" align="center"></p>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p>see also here for the R package pages <a href="https://github.com/kullrich/scTEI" class="external-link">https://github.com/kullrich/scTEI</a></p>
<div class="section level2">
<h2 id="r-specific-installation-prerequisites">R specific installation prerequisites<a class="anchor" aria-label="anchor" href="#r-specific-installation-prerequisites"></a>
</h2>
<div class="section level3">
<h3 id="install-packages-from-cran">install packages from <a href="https://cran.r-project.org/web/packages/index.html" class="external-link">cran</a>
<a class="anchor" aria-label="anchor" href="#install-packages-from-cran"></a>
</h3>
<p>In most cases you need to first install the following system-wide packages to be able to compile the R dependencies.</p>
<p>Ubuntu/Debian</p>
<pre><code>sudo apt-get install libcurl4-openssl-dev libssl-dev libxml2-dev libglu1-mesa-dev libgit2-dev
#pkgdown dependencies - pkgdown is used to build R package pages
#sudo apt-get install libssh2-1-dev libfontconfig1-dev libharfbuzz-dev libfribidi-dev</code></pre>
<p>CentOS</p>
<pre><code>sudo yum install libcurl-devel openssl-devel libxml2-devel mesa-libGLU-devel libgit2-devel
#pkgdown dependencies - pkgdown is used to build R package pages
#sudo yum install libssh2-devel fontconfig-devel harfbuzz-devel fribidi-devel</code></pre>
<ul>
<li><a href="https://cran.r-project.org/web/packages/Rcpp/index.html" class="external-link">Rcpp</a></li>
<li><a href="https://cran.r-project.org/web/packages/RcppArmadillo/index.html" class="external-link">RcppArmadillo</a></li>
<li><a href="https://cran.r-project.org/web/packages/RcppThread/index.html" class="external-link">RcppThread</a></li>
<li><a href="https://cran.r-project.org/web/packages/devtools/index.html" class="external-link">devtools</a></li>
<li><a href="https://cran.r-project.org/web/packages/testthat/index.html" class="external-link">testthat</a></li>
<li><a href="https://cran.r-project.org/web/packages/dplyr/index.html" class="external-link">dplyr</a></li>
<li><a href="https://cran.r-project.org/web/packages/plyr/index.html" class="external-link">plyr</a></li>
<li><a href="https://cran.r-project.org/web/packages/ggplot2/index.html" class="external-link">ggplot2</a></li>
<li><a href="https://cran.r-project.org/web/packages/viridis/index.html" class="external-link">viridis</a></li>
<li><a href="https://cran.r-project.org/web/packages/cowplot/index.html" class="external-link">cowplot</a></li>
<li><a href="https://cran.r-project.org/web/packages/Seurat/index.html" class="external-link">Seurat</a></li>
<li><a href="https://cran.r-project.org/web/packages/SeuratData/index.html" class="external-link">SeuratData</a></li>
<li><a href="https://cran.r-project.org/web/packages/Matrix.utils/index.html" class="external-link">Matrix.utils</a></li>
<li><a href="https://cran.r-project.org/web/packages/myTAI/index.html" class="external-link">myTAI</a></li>
</ul>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"Rcpp"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"RcppArmadillo"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"RcppThread"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"testthat"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"plyr"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"viridis"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"cowplot"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"Seurat"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"SeuratData"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"Matrix.utils"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"myTAI"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="install-packages-from-bioconductor">install packages from <a href="https://www.bioconductor.org/" class="external-link">Bioconductor</a>
<a class="anchor" aria-label="anchor" href="#install-packages-from-bioconductor"></a>
</h3>
<ul>
<li><a href="https://bioconductor.org/packages/release/bioc/html/BiocGenerics.html" class="external-link">BiocGenerics</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/DelayedArray.html" class="external-link">DelayedArray</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/DelayedMatrixStats.html" class="external-link">DelayedMatrixStats</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/limma.html" class="external-link">limma</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/S4Vectors.html" class="external-link">S4Vectors</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html" class="external-link">SummarizedExperiment</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/batchelor.html" class="external-link">batchelor</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html" class="external-link">ComplexHeatmap</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/org.Hs.eg.db.html" class="external-link">org.Hs.eg.db</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/org.Mm.eg.db.html" class="external-link">org.Mm.eg.db</a></li>
</ul>
<pre><code><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"BiocGenerics"</span>,</span>
<span>    <span class="st">"DelayedArray"</span>,</span>
<span>    <span class="st">"DelayedMatrixStats"</span>,</span>
<span>    <span class="st">"limma"</span>,</span>
<span>    <span class="st">"S4Vectors"</span>,</span>
<span>    <span class="st">"SingleCellExperiment"</span>,</span>
<span>    <span class="st">"SummarizedExperiment"</span>,</span>
<span>    <span class="st">"batchelor"</span>,</span>
<span>    <span class="st">"ComplexHeatmap"</span>,</span>
<span>    <span class="st">"org.Hs.eg.db"</span>,</span>
<span>    <span class="st">"org.Mm.eg.db"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="install-monocle3">install <a href="https://cole-trapnell-lab.github.io/monocle3/docs/installation/" class="external-link">monocle3</a>
<a class="anchor" aria-label="anchor" href="#install-monocle3"></a>
</h3>
<pre><code><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"cole-trapnell-lab/leidenbase"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"cole-trapnell-lab/monocle3"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="install-sctei">install <a href="https://github.com/kullrich/scTEI" class="external-link">scTEI</a>
<a class="anchor" aria-label="anchor" href="#install-sctei"></a>
</h3>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"kullrich/scTEI"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="quick-guide">Quick-guide<a class="anchor" aria-label="anchor" href="#quick-guide"></a>
</h2>
<div class="section level3">
<h3 id="seurat-object">Seurat object<a class="anchor" aria-label="anchor" href="#seurat-object"></a>
</h3>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kullrich/scTEI" class="external-link">scTEI</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## get Seurat object</span></span>
<span><span class="fu">SeuratData</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html" class="external-link">InstallData</a></span><span class="op">(</span><span class="st">"celegans.embryo.SeuratData"</span><span class="op">)</span></span>
<span><span class="va">celegans</span> <span class="op">&lt;-</span> <span class="fu">SeuratData</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span><span class="st">"celegans.embryo"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## load Caenorhabditis elegans gene age estimation</span></span>
<span><span class="va">celegans_ps</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span></span>
<span>    file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>    <span class="st">"Sun2021_Orthomap.tsv"</span>, package <span class="op">=</span> <span class="st">"scTEI"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">## define Phylostratum</span></span>
<span><span class="va">ps_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">celegans_ps</span><span class="op">$</span><span class="va">Phylostratum</span><span class="op">)</span>,</span>
<span>    <span class="va">celegans_ps</span><span class="op">$</span><span class="va">GeneID</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## add TEI values</span></span>
<span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span><span class="st">"TEI"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/TEI.html">TEI</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">celegans</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Use multiple threads to calculate TEI on sparseMatrix</span></span>
<span><span class="co">#celegans@meta.data["TEI"] &lt;- TEI(</span></span>
<span><span class="co">#    ExpressionSet = celegans@assays$RNA@counts,</span></span>
<span><span class="co">#    Phylostratum = ps_vec,</span></span>
<span><span class="co">#    split = 1000,</span></span>
<span><span class="co">#    threads = 2</span></span>
<span><span class="co">#)</span></span>
<span></span>
<span><span class="co">## make RidgePlot</span></span>
<span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"embryo.time.bin"</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RidgePlot.html" class="external-link">RidgePlot</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">celegans</span>,</span>
<span>    features <span class="op">=</span> <span class="st">"TEI"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p1</span></span>
<span></span>
<span><span class="co">## preprocess scRNA</span></span>
<span><span class="va">all.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">)</span></span>
<span><span class="va">celegans</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span></span>
<span>    <span class="va">celegans</span>,</span>
<span>    normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span>,</span>
<span>    scale.factor <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>selection.method <span class="op">=</span> <span class="st">"vst"</span>,</span>
<span>    nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>features <span class="op">=</span> <span class="va">all.genes</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>dims<span class="op">=</span><span class="fl">50</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## make FeaturePlot</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">celegans</span><span class="op">)</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">FeaturePlot</span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">celegans</span>,</span>
<span>    features <span class="op">=</span> <span class="st">"TEI"</span>,</span>
<span>    min.cutoff<span class="op">=</span><span class="st">'q05'</span>,</span>
<span>    max.cutoff<span class="op">=</span><span class="st">'q95'</span>,</span>
<span>    cols <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p2</span>, <span class="va">p3</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="monocle3-object">Monocle3 object<a class="anchor" aria-label="anchor" href="#monocle3-object"></a>
</h3>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kullrich/scTEI" class="external-link">scTEI</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## get Monocle3 object</span></span>
<span><span class="va">expression_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"http://staff.washington.edu/hpliner/data/"</span>,</span>
<span>    <span class="st">"packer_embryo_expression.rds"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">cell_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"http://staff.washington.edu/hpliner/data/"</span>,</span>
<span>    <span class="st">"packer_embryo_colData.rds"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">gene_annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"http://staff.washington.edu/hpliner/data/"</span>,</span>
<span>    <span class="st">"packer_embryo_rowData.rds"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/new_cell_data_set.html" class="external-link">new_cell_data_set</a></span><span class="op">(</span></span>
<span>    expression_data <span class="op">=</span> <span class="va">expression_matrix</span>,</span>
<span>    cell_metadata <span class="op">=</span> <span class="va">cell_metadata</span>,</span>
<span>    gene_metadata <span class="op">=</span> <span class="va">gene_annotation</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## load Caenorhabditis elegans gene age estimation</span></span>
<span><span class="va">celegans_ps</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span></span>
<span>    file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>    <span class="st">"Sun2021_Orthomap.tsv"</span>, package <span class="op">=</span> <span class="st">"scTEI"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## define Phylostratum</span></span>
<span><span class="va">ps_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">celegans_ps</span><span class="op">$</span><span class="va">Phylostratum</span><span class="op">)</span>,</span>
<span>    <span class="va">celegans_ps</span><span class="op">$</span><span class="va">GeneID</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## add TEI values</span></span>
<span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">cds</span><span class="op">)</span><span class="op">[</span><span class="st">"TEI"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/TEI.html">TEI</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="fu">SingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">cds</span><span class="op">)</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Use multiple threads to calculate TEI on sparseMatrix</span></span>
<span><span class="co">#SummarizedExperiment::colData(cds)["TEI"] &lt;- TEI(</span></span>
<span><span class="co">#    ExpressionSet = SingleCellExperiment::counts(cds),</span></span>
<span><span class="co">#    Phylostratum = ps_vec,</span></span>
<span><span class="co">#    split = 1000,</span></span>
<span><span class="co">#    threads = 2</span></span>
<span><span class="co">#)</span></span>
<span></span>
<span><span class="co">## make boxplot by embryo.time</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">cds</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">embryo.time.bin</span>, y<span class="op">=</span><span class="va">TEI</span>, fill<span class="op">=</span><span class="va">embryo.time.bin</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html" class="external-link">geom_violin</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>width<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="va">p1</span></span>
<span></span>
<span><span class="co">## preprocess scRNA</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/preprocess_cds.html" class="external-link">preprocess_cds</a></span><span class="op">(</span><span class="va">cds</span>, num_dim <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/align_cds.html" class="external-link">align_cds</a></span><span class="op">(</span><span class="va">cds</span>, alignment_group <span class="op">=</span> <span class="st">"batch"</span>,</span>
<span>    residual_model_formula_str <span class="op">=</span> <span class="st">"~ bg.300.loading +</span></span>
<span><span class="st">    bg.400.loading + bg.500.1.loading + bg.500.2.loading +</span></span>
<span><span class="st">    bg.r17.loading + bg.b01.loading + bg.b02.loading"</span><span class="op">)</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/reduce_dimension.html" class="external-link">reduce_dimension</a></span><span class="op">(</span><span class="va">cds</span><span class="op">)</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/cluster_cells.html" class="external-link">cluster_cells</a></span><span class="op">(</span><span class="va">cds</span><span class="op">)</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/learn_graph.html" class="external-link">learn_graph</a></span><span class="op">(</span><span class="va">cds</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## make FeaturePlot</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/plot_cells.html" class="external-link">plot_cells</a></span><span class="op">(</span><span class="va">cds</span>,</span>
<span>    label_groups_by_cluster<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>    color_cells_by <span class="op">=</span> <span class="st">"embryo.time.bin"</span>,</span>
<span>    group_label_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    label_cell_groups <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    label_leaves <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    label_branch_points <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    graph_label_size<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">monocle3</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/plot_cells.html" class="external-link">plot_cells</a></span><span class="op">(</span><span class="va">cds</span>,</span>
<span>    label_groups_by_cluster<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>    color_cells_by <span class="op">=</span> <span class="st">"TEI"</span>,</span>
<span>    group_label_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    label_cell_groups <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    label_leaves <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    label_branch_points <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    graph_label_size<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>    cell_size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p2</span>, <span class="va">p3</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="some-examples">Some examples<a class="anchor" aria-label="anchor" href="#some-examples"></a>
</h2>
<div class="section level3">
<h3 id="caenorhabditis-elegans-packer-and-zhu-et-al-2019-sun-et-al-2021">Caenorhabditis elegans (Packer and Zhu et al. 2019, Sun et al. 2021)<a class="anchor" aria-label="anchor" href="#caenorhabditis-elegans-packer-and-zhu-et-al-2019-sun-et-al-2021"></a>
</h3>
<p>single-cell data: Packer and Zhu et al. 2019</p>
<p>ortho map: Sun et al. 2021</p>
<p><img src="reference/figures/celegans_seurat_p1.png" align="center"></p>
<p><img src="reference/figures/celegans_monocle3_p1.png" align="center"></p>
<p><img src="reference/figures/celegans_monocle3_p2_p3.png" align="center"></p>
</div>
<div class="section level3">
<h3 id="hydra-vulgaris-cazet-et-al-2022">Hydra vulgaris (Cazet et al. 2022)<a class="anchor" aria-label="anchor" href="#hydra-vulgaris-cazet-et-al-2022"></a>
</h3>
<p>single-cell data: Cazet et al. 2022</p>
<p>ortho map: Cazet et al. 2022</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kullrich/scTEI" class="external-link">scTEI</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## increase timeout - since download of 1.9 GB file</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout<span class="op">=</span><span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## get Seurat object</span></span>
<span><span class="va">hvulgaris</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://research.nhgri.nih.gov/"</span>,</span>
<span>    <span class="st">"HydraAEP/download/scriptsdata/"</span>,</span>
<span>    <span class="st">"aepAtlasNonDub.rds"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## re-order meta.data according to cell order</span></span>
<span><span class="va">hvulgaris</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="va">hvulgaris</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">hvulgaris</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">SCT</span><span class="op">@</span><span class="va">data</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">hvulgaris</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co">## load Hydra vulgaris gene age estimation</span></span>
<span><span class="va">hvulgaris_ps</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span></span>
<span>    file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>    <span class="st">"Cazet2022_Orthomap.tsv"</span>, package <span class="op">=</span> <span class="st">"scTEI"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span>    </span>
<span><span class="co">## define Phylostratum</span></span>
<span><span class="va">ps_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">hvulgaris_ps</span><span class="op">$</span><span class="va">ageN</span><span class="op">)</span>,</span>
<span>    <span class="va">hvulgaris_ps</span><span class="op">$</span><span class="va">ID</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## add TEI values</span></span>
<span><span class="va">hvulgaris</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span><span class="st">"TEI"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/TEI.html">TEI</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">hvulgaris</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">SCT</span><span class="op">@</span><span class="va">data</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span>,</span>
<span>    split <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    threads <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## make RidgePlot</span></span>
<span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">hvulgaris</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"curatedIdent"</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RidgePlot.html" class="external-link">RidgePlot</a></span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">hvulgaris</span>,</span>
<span>    features <span class="op">=</span> <span class="st">"TEI"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p1</span></span>
<span></span>
<span><span class="co">## make FeaturePlot</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">hvulgaris</span><span class="op">)</span></span>
<span><span class="va">p2</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">FeaturePlot</span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">hvulgaris</span>,</span>
<span>    features <span class="op">=</span> <span class="st">"TEI"</span>,</span>
<span>    min.cutoff<span class="op">=</span><span class="st">'q05'</span>,</span>
<span>    max.cutoff<span class="op">=</span><span class="st">'q95'</span>,</span>
<span>    cols <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p3</span></span>
<span></span>
<span><span class="co">## get relative expression per cell group</span></span>
<span><span class="va">cell_groups</span> <span class="op">&lt;-</span> <span class="fu">Ident2cellList</span><span class="op">(</span><span class="fu">Idents</span><span class="op">(</span><span class="va">hvulgaris</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">reM</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/REMatrix.html">REMatrix</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">hvulgaris</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">SCT</span><span class="op">@</span><span class="va">data</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span>,</span>
<span>    groups <span class="op">=</span> <span class="va">cell_groups</span>,</span>
<span>    by <span class="op">=</span> <span class="st">"row"</span>,</span>
<span>    threads <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span></span>
<span>    <span class="va">reM</span>,</span>
<span>    name <span class="op">=</span> <span class="st">"RExp"</span>,</span>
<span>    column_title <span class="op">=</span> <span class="st">"Relative Expression Profile - cell groups"</span>,</span>
<span>    row_title <span class="op">=</span> <span class="st">"More Recent &lt;&lt;&lt; More Ancient"</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    cluster_columns <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    col <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p4</span></span></code></pre>
<p><img src="reference/figures/hvulgaris_seurat_p1.png" align="center"></p>
<p><img src="reference/figures/hvulgaris_seurat_p2.png" align="center"></p>
<p><img src="reference/figures/hvulgaris_seurat_p3.png" align="center"></p>
<p><img src="reference/figures/hvulgaris_seurat_p4.png" align="center"></p>
</div>
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p>GPL-3 (see LICENSE)</p>
</div>
<div class="section level2">
<h2 id="contributing-code">Contributing Code<a class="anchor" aria-label="anchor" href="#contributing-code"></a>
</h2>
<p>If you would like to contribute to scTEI, please file an issue so that one can establish a statement of need, avoid redundant work, and track progress on your contribution.</p>
<p>Before you do a pull request, you should always file an issue and make sure that someone from the scTEI developer team agrees that it’s a problem, and is happy with your basic proposal for fixing it.</p>
<p>Once an issue has been filed and we’ve identified how to best orient your contribution with package development as a whole, <a href="https://docs.github.com/en/github/getting-started-with-github/fork-a-repo" class="external-link">fork</a> the <a href="https://github.com/kullrich/scTEI.git" class="external-link">main repo</a>, branch off a <a href="https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-branches" class="external-link">feature branch</a> from <code>main</code>, <a href="https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/committing-and-reviewing-changes-to-your-project" class="external-link">commit</a> and <a href="https://docs.github.com/en/github/using-git/pushing-commits-to-a-remote-repository" class="external-link">push</a> your changes to your fork and submit a <a href="https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/proposing-changes-to-your-work-with-pull-requests" class="external-link">pull request</a> for <code>scTEI:main</code>.</p>
<p>By contributing to this project, you agree to abide by the Code of Conduct terms.</p>
</div>
<div class="section level2">
<h2 id="bug-reports">Bug reports<a class="anchor" aria-label="anchor" href="#bug-reports"></a>
</h2>
<p>Please report any errors or requests regarding <a href="https://github.com/kullrich/scTEI" class="external-link">scTEI</a> to Kristian Ullrich (<a href="mailto:ullrich@evolbio.mpg.de" class="email">ullrich@evolbio.mpg.de</a>)</p>
<p>or use the issue tracker at <a href="https://github.com/kullrich/scTEI/issues" class="external-link">https://github.com/kullrich/scTEI/issues</a></p>
</div>
<div class="section level2">
<h2 id="code-of-conduct---participation-guidelines">Code of Conduct - Participation guidelines<a class="anchor" aria-label="anchor" href="#code-of-conduct---participation-guidelines"></a>
</h2>
<p>This repository adhere to <a href="https://contributor-covenant.org" class="external-link">Contributor Covenant</a> code of conduct for in any interactions you have within this project. (see <a href="https://github.com/kullrich/scTEI/-/blob/main/CODE_OF_CONDUCT.md" class="external-link">Code of Conduct</a>)</p>
<p>See also the policy against sexualized discrimination, harassment and violence for the Max Planck Society <a href="https://www.mpg.de/11961177/code-of-conduct-en.pdf" class="external-link">Code-of-Conduct</a>.</p>
<p>By contributing to this project, you agree to abide by its terms.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Cao, Junyue, Malte Spielmann, Xiaojie Qiu, Xingfan Huang, Daniel M. Ibrahim, Andrew J. Hill, Fan Zhang et al. (2019) <strong>The single-cell transcriptional landscape of mammalian organogenesis.</strong> <em>Nature</em>, <strong>566, no. 7745</strong>. <a href="https://doi.org/10.1038/s41586-019-0969-x" class="external-link">https://doi.org/10.1038/s41586-019-0969-x</a></p>
<p>Cazet, Jack, Stefan Siebert, Hannah Morris Little, Philip Bertemes, Abby S. Primack, Peter Ladurner, Matthias Achrainer et al. (2022) <strong>New Hydra genomes reveal conserved principles of hydrozoan transcriptional regulation.</strong>, <em>bioRxiv</em>, <strong>2022.06.21.496857</strong>. <a href="https://doi.org/10.1101/2022.06.21.496857" class="external-link">https://doi.org/10.1101/2022.06.21.496857</a></p>
<p>Domazet-Lošo, Tomislav, Josip Brajković, and Diethard Tautz. (2007) <strong>A Phylostratigraphy Approach to Uncover the Genomic History of Major Adaptations in Metazoan Lineages.</strong> <em>Trends in Genetics</em>, <strong>23, no.11</strong>. <a href="https://doi.org/10.1016/j.tig.2007.08.014" class="external-link">https://doi.org/10.1016/j.tig.2007.08.014</a></p>
<p>Domazet-Lošo, Tomislav, and Diethard Tautz. (2010) <strong>A phylogenetically based transcriptome age index mirrors ontogenetic divergence patterns.</strong> <em>Nature</em>, <strong>468, no. 7325</strong>. <a href="https://doi.org/10.1038/nature09632" class="external-link">https://doi.org/10.1038/nature09632</a></p>
<p>Drost, Hajk-Georg, Alexander Gabel, Ivo Grosse, and Marcel Quint. (2015) <strong>Evidence for active maintenance of phylotranscriptomic hourglass patterns in animal and plant embryogenesis.</strong> <em>Molecular biology and evolution</em>, <strong>32, no. 5</strong>. <a href="https://doi.org/10.1093/molbev/msv012" class="external-link">https://doi.org/10.1093/molbev/msv012</a></p>
<p>Emms, David M, and Steven Kelly. (2019) <strong>OrthoFinder: Phylogenetic Orthology Inference for Comparative Genomics.</strong> <em>Genome Biology</em>, <strong>20, no. 1</strong>. <a href="https://doi.org/10.1186/s13059-019-1832-y" class="external-link">https://doi.org/10.1186/s13059-019-1832-y</a></p>
<p>Gossmann, Toni I, Dounia Saleh, Marc W Schmid, Michael A Spence, and Karl J Schmid. (2016) <strong>Transcriptomes of Plant Gametophytes Have a Higher Proportion of Rapidly Evolving and Young Genes Than Sporophytes.</strong> <em>Molecular Biology and Evolution</em>, <strong>33, no. 7</strong>. <a href="https://doi.org/10.1093/molbev/msw044" class="external-link">https://doi.org/10.1093/molbev/msw044</a></p>
<p>Julca, Irene, Camilla Ferrari, Marı́a Flores-Tornero, Sebastian Proost, Ann-Cathrin Lindner, Dieter Hackenberg, Lenka Steinbachová, et al. (2021) <strong>Comparative Transcriptomic Analysis Reveals Conserved Programmes Underpinning Organogenesis and Reproduction in Land Plants.</strong> <em>Nature Plants</em>, <strong>7, no. 8</strong>. <a href="https://doi.org/10.1038/s41477-021-00958-2" class="external-link">https://doi.org/10.1038/s41477-021-00958-2</a></p>
<p>Lechner, Marcus, Sven Findeiß, Lydia Steiner, Manja Marz, Peter F Stadler, and Sonja J Prohaska. (2011) <strong>Proteinortho: Detection of (Co-) Orthologs in Large-Scale Analysis.</strong> <em>BMC Bioinformatics</em>, <strong>12, no. 1</strong>. <a href="https://doi.org/10.1186/1471-2105-12-124" class="external-link">https://doi.org/10.1186/1471-2105-12-124</a></p>
<p>Linard, Benjamin, Ingo Ebersberger, Shawn E McGlynn, Natasha Glover, Tomohiro Mochizuki, Mateus Patricio, Odile Lecompte, et al. (2021) <strong>Ten Years of Collaborative Progress in the Quest for Orthologs.</strong> <em>Molecular Biology and Evolution</em>, <strong>38, no. 8</strong>. <a href="https://doi.org/10.1093/molbev/msab098" class="external-link">https://doi.org/10.1093/molbev/msab098</a></p>
<p>Liu, Jialin, Michael Frochaux, Vincent Gardeux, Bart Deplancke, and Marc Robinson-Rechavi. (2020) <strong>Inter-Embryo Gene Expression Variability Recapitulates the Hourglass Pattern of Evo-Devo.</strong> <em>BMC Biology</em>, <strong>18, no. 1</strong>. <a href="https://doi.org/10.1186/s12915-020-00842-z" class="external-link">https://doi.org/10.1186/s12915-020-00842-z</a></p>
<p>Packer, Jonathan S., Qin Zhu, Chau Huynh, Priya Sivaramakrishnan, Elicia Preston, Hannah Dueck, Derek Stefanik et al. (2019) <strong>A lineage-resolved molecular atlas of C. elegans embryogenesis at single-cell resolution.</strong> <em>Science</em>, <strong>365, no. 6459</strong>. <a href="https://doi.org/10.1126/science.aax1971" class="external-link">https://doi.org/10.1126/science.aax1971</a></p>
<p>Proost, Sebastian, Michiel Van Bel, Lieven Sterck, Kenny Billiau, Thomas Van Parys, Yves Van de Peer, and Klaas Vandepoele. (2009) <strong>PLAZA: A Comparative Genomics Resource to Study Gene and Genome Evolution in Plants.</strong> <em>The Plant Cell</em>, <strong>21 no. 12</strong>. <a href="https://doi.org/10.1105/tpc.109.071506" class="external-link">https://doi.org/10.1105/tpc.109.071506</a></p>
<p>Quint, Marcel, Hajk-Georg Drost, Alexander Gabel, Kristian Karsten Ullrich, Markus Bönn, and Ivo Grosse. (2012) <strong>A transcriptomic hourglass in plant embryogenesis.</strong> <em>Nature</em>, <strong>490, no. 7418</strong>. <a href="https://doi.org/10.1038/nature11394" class="external-link">https://doi.org/10.1038/nature11394</a></p>
<p>Satija, Rahul, Jeffrey A. Farrell, David Gennert, Alexander F. Schier, and Aviv Regev. (2015) <strong>Spatial reconstruction of single-cell gene expression data.</strong> <em>Nature biotechnology</em>, <strong>33, no. 5</strong>. <a href="https://doi.org/10.1038/nbt.3192" class="external-link">https://doi.org/10.1038/nbt.3192</a></p>
<p>Schneider, Adrian, Christophe Dessimoz, and Gaston H Gonnet. (2007) <strong>OMA Browser—Exploring Orthologous Relations Across 352 Complete Genomes.</strong> <em>Bioinformatics</em>, <strong>23, no. 16</strong>. <a href="https://doi.org/10.1093/bioinformatics/btm295" class="external-link">https://doi.org/10.1093/bioinformatics/btm295</a></p>
<p>Sun, Shuai, Christian Rödelsperger, and Ralf J Sommer. (2021) <strong>Single Worm Transcriptomics Identifies a Developmental Core Network of Oscillating Genes with Deep Conservation Across Nematodes.</strong> <em>Genome Research</em>, <strong>31, no. 9</strong>. <a href="https://doi.org/10.1101/gr.275303.121" class="external-link">https://doi.org/10.1101/gr.275303.121</a></p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/kullrich/scTEI/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/kullrich/scTEI/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scTEI</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Kristian K Ullrich <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-4308-9626" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Kristian K Ullrich.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
