<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single-Cell Expression Analysis with `scTEI` • scTEI</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Single-Cell Expression Analysis with `scTEI`">
<meta property="og:description" content="scTEI">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scTEI</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/scTEI.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kullrich/scTEI/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Single-Cell Expression Analysis with
<code>scTEI</code>
</h1>
                        <h4 data-toc-skip class="author">Kristian K
Ullrich</h4>
            
            <h4 data-toc-skip class="date">2022-10-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/kullrich/scTEI/blob/HEAD/vignettes/scTEI.Rmd" class="external-link"><code>vignettes/scTEI.Rmd</code></a></small>
      <div class="hidden name"><code>scTEI.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>scTEI add any phylogenetically based transcriptome evolutionary
      index (TEI) to single-cell data objects</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The goal of <code>scTEI</code> is to provide easy functionality to
add phylogenetically based transcriptome evolutionary index (TEI) to
single-cell data objects.</p>
<p>For a comprehensive overview about the topic of gene age assignments,
transcriptome age index (TAI) and its derivates (TDI, TPI, Adjusted SD,
PhastCons, …) see e.g.:</p>
<ul>
<li><a href="https://drostlab.github.io/myTAI/articles/Introduction.html" class="external-link">myTAI
- Introduction</a></li>
<li>TAI: <span class="citation">(Domazet-Lošo, Brajković, and Tautz
2007)</span>
</li>
<li>TAI: <span class="citation">(Domazet-Lošo and Tautz
2010)</span>
</li>
<li>TDI: <span class="citation">(Quint et al. 2012)</span>
</li>
<li>TDI: <span class="citation">(Drost et al. 2017)</span>
</li>
<li>TPI: <span class="citation">(Gossmann et al. 2016)</span>
</li>
<li>Adjusted SD, PhastCons: <span class="citation">(Liu et al.
2020)</span>
</li>
</ul>
<p>Since all of these values deal with the topic of weigthing
transcriptome data with an evolutionary index, I combine all of them
under one term <em>transcriptome evolutionary index</em>, short
<em>TEI</em>.</p>
<p>The following sections introduce main <em>single-cell</em> gene
expression data analysis techniques implemented in <code>scTEI</code>,
which basically update basic <code>myTAI</code> functions to deal with
large <em>single-cell</em> data objects:</p>
<ul>
<li><p>Adding transcriptome evolutionary index (TEI) to single-cell data
objects</p></li>
<li><p>Seurat</p></li>
<li><p>monocle3</p></li>
<li><p>Visualizing TEI values</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="adding-transcriptome-evolutionary-index-tei-to-single-cell-data-objects">Adding transcriptome evolutionary index (TEI) to single-cell data
objects<a class="anchor" aria-label="anchor" href="#adding-transcriptome-evolutionary-index-tei-to-single-cell-data-objects"></a>
</h2>
<p>A variety of data objects are used to deal with sparse count data
(<code>dgCMatrix</code>) from single-cell RNA sequencing
(scRNA-seq).</p>
<p>Here, we focus on the common used R packages <code>Seurat</code> and
<code>monocle3</code>, which both use a <code>sparseMatrix</code> object
to store scRNA count data.</p>
<p><strong>Note:</strong> Apparently, one important note is that count
data or subsequent normalized counts or scaled data needs to be
positive.</p>
<p>One should consider transforming scaled data to positive scale prior
applying TEI calculations.</p>
<p>In this section we will introduce how to add TEI values to either a
<code>seurat</code> or <code>monocle3</code> cell data set.</p>
<p><strong>Note that prior calculating TEI one needs to retrieve
phylogenetic or taxonomic information for your focal
species.</strong></p>
<p>This might be a <em>phylostratigraphic map</em> (as introduced by
<span class="citation">(Domazet-Lošo, Brajković, and Tautz 2007)</span>)
or an <em>ortho map</em> (I call them, see e.g. <span class="citation">(Julca et al. 2021)</span> or <span class="citation">(Cazet et al. 2022)</span>), which can be obtained by
assigning to each orthogroup (OG) or hierachical orthogroup (HOG) along
a given species tree the ancestral node.</p>
<p>Please have a look at the introduction of the great
<code>myTAI</code> package for possible sources, how to get such
<em>phylostratigraphic maps</em> (<a href="https://drostlab.github.io/myTAI/articles/Introduction.html" class="external-link">myTAI
- Introduction</a>).</p>
<p>To create an <em>ortho map</em>, one simply needs to get OGs or HOGs
with e.g. OrthoFinder (<span class="citation">(Emms and Kelly
2019)</span>) or Proteinortho (<span class="citation">(Lechner et al.
2011)</span>) or any other ortholog prediction tool (see <span class="citation">(Linard et al. 2021)</span>) using a set of species
that cover the species range of your interest. Parse each OG or HOG for
the <em>oldest clade</em> as compared to a species tree and your focal
species of interest.</p>
<p>Or e.g. use pre-calculated OGs from e.g. <a href="https://omabrowser.org/oma/home/" class="external-link">https://omabrowser.org/oma/home</a>
(<span class="citation">(Schneider, Dessimoz, and Gonnet 2007)</span>)
or <a href="https://bioinformatics.psb.ugent.be/plaza/" class="external-link">https://bioinformatics.psb.ugent.be/plaza/</a>
for plants (<span class="citation">(Proost et al. 2009)</span>).</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="install---seurat-and-monocle3">Install - Seurat and monocle3<a class="anchor" aria-label="anchor" href="#install---seurat-and-monocle3"></a>
</h3>
<p>Please visit the following documentation to install
<code>Seurat</code> and <code>monocole3</code></p>
<p><a href="https://satijalab.org/seurat/articles/install.html" class="external-link">Seurat</a></p>
<p><a href="https://cole-trapnell-lab.github.io/monocle3/docs/installation/" class="external-link">monocle3</a></p>
</div>
<div class="section level3">
<h3 id="install-org-hs-eg-db-and-org-mm-eg-db-to-be-able-to-convert-ensembl-ids-into-gene-name-alias">Install <code>org.Hs.eg.db</code> and <code>org.Mm.eg.db</code> to
be able to convert ensembl IDs into gene name alias<a class="anchor" aria-label="anchor" href="#install-org-hs-eg-db-and-org-mm-eg-db-to-be-able-to-convert-ensembl-ids-into-gene-name-alias"></a>
</h3>
<p><a href="https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html" class="external-link">org.Hs.eg.db</a>
<a href="https://bioconductor.org/packages/release/data/annotation/html/org.Mm.eg.db.html" class="external-link">org.Mm.eg.db</a></p>
<pre><code><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"org.Hs.eg.db"</span>,</span>
<span>    <span class="st">"org.Mm.eg.db"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="install-additional-plotting-packages">Install additional plotting packages<a class="anchor" aria-label="anchor" href="#install-additional-plotting-packages"></a>
</h3>
<p><a href="https://cran.r-project.org/web/packages/ggplot2/vignettes/ggplot2-in-packages.html" class="external-link">ggplot2</a>
<a href="https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html" class="external-link">viridis</a>
<a href="https://cran.r-project.org/web/packages/cowplot/vignettes/introduction.html" class="external-link">cowplot</a>
<a href="https://bioconductor.org/packages/release/bioc/vignettes/ComplexHeatmap/inst/doc/complex_heatmap.html" class="external-link">ComplexHeatmap</a></p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"viridis"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"cowplot"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ComplexHeatmap"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="load-libraries">Load libraries<a class="anchor" aria-label="anchor" href="#load-libraries"></a>
</h3>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kullrich/scTEI" class="external-link">scTEI</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat" class="external-link">SeuratData</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">monocle3</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="seurat---example">Seurat - example<a class="anchor" aria-label="anchor" href="#seurat---example"></a>
</h2>
<p>For <code>Seurat</code>, we use scRNA data obtained from
<code>SeuratData</code> for the model organism <em>Caenorhabditis
elegans</em> <span class="citation">(Packer et al. 2019)</span> and
phylogenetic data obtained from <span class="citation">(Sun,
Rödelsperger, and Sommer 2021)</span>
(Sun2021_Supplemental_Table_S6).</p>
<div class="section level3">
<h3 id="example-6k-c--elegans-embryos-from-packer-and-zhu-et-al-2019">Example: 6k C. elegans embryos from Packer and Zhu et al (2019)<a class="anchor" aria-label="anchor" href="#example-6k-c--elegans-embryos-from-packer-and-zhu-et-al-2019"></a>
</h3>
<p>Here, for demonstration purposes, we will be using the 6k
<em>Caenorhabditis elegans</em> embryos object that is available via the
<code>SeuratData</code> package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load Packer and Zhu et al (2019) data set</span></span>
<span><span class="fu">SeuratData</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html" class="external-link">InstallData</a></span><span class="op">(</span><span class="st">"celegans.embryo.SeuratData"</span><span class="op">)</span></span>
<span><span class="va">celegans</span> <span class="op">&lt;-</span> <span class="fu">SeuratData</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span><span class="st">"celegans.embryo"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 20222  6188</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "WBGene00010957" "WBGene00010958" "WBGene00010959" "WBGene00010960"</span></span>
<span><span class="co">## [5] "WBGene00010961" "WBGene00000829"</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># preprocess scRNA</span></span>
<span><span class="va">all.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">)</span></span>
<span><span class="va">celegans</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span></span>
<span>    <span class="va">celegans</span>,</span>
<span>    normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span>,</span>
<span>    scale.factor <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>selection.method <span class="op">=</span> <span class="st">"vst"</span>,</span>
<span>    nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>features <span class="op">=</span> <span class="va">all.genes</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>dims<span class="op">=</span><span class="fl">50</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"embryo.time.bin"</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">celegans</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p1</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># overwrite NA in cell.type + add embryo.time.bin x cell.type</span></span>
<span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">cell.type</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">cell.type</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"notClassified"</span></span>
<span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span><span class="st">"embryo.time.bin.cell.type"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span><span class="st">"embryo.time.bin"</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    <span class="st">"-"</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span><span class="st">"cell.type"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load Caenorhabditis elegans gene age estimation</span></span>
<span><span class="va">celegans_ps</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>    <span class="st">"Sun2021_Orthomap.tsv"</span>, package <span class="op">=</span> <span class="st">"scTEI"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">celegans_ps</span><span class="op">$</span><span class="va">Phylostratum</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##    0    1    2    3    4    5    6    7    8    9   10   11   12   13 </span></span>
<span><span class="co">## 1290 5434 4666  603 1039  808  274  884  590  511  384 1113 1277 1167</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># define Phylostratum</span></span>
<span><span class="va">ps_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">celegans_ps</span><span class="op">$</span><span class="va">Phylostratum</span><span class="op">)</span>,</span>
<span>    <span class="va">celegans_ps</span><span class="op">$</span><span class="va">GeneID</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add TEI values</span></span>
<span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span><span class="st">"TEI"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TEI.html">TEI</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">celegans</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Use multiple threads to calculate TEI on sparseMatrix</span></span>
<span></span>
<span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span><span class="st">"TEI"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TEI.html">TEI</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">celegans</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span>,</span>
<span>    split <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make FeaturePlot</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">FeaturePlot</span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">celegans</span>,</span>
<span>    features <span class="op">=</span> <span class="st">"TEI"</span>,</span>
<span>    min.cutoff<span class="op">=</span><span class="st">'q05'</span>,</span>
<span>    max.cutoff<span class="op">=</span><span class="st">'q95'</span>,</span>
<span>    cols <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p2</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make RidgePlot</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">RidgePlot</span><span class="op">(</span>object <span class="op">=</span> <span class="va">celegans</span>,</span>
<span>    features <span class="op">=</span> <span class="st">"TEI"</span>,</span>
<span>    group.by <span class="op">=</span> <span class="st">"embryo.time.bin"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Picking joint bandwidth of 0.0512</span></span></code></pre>
<p><img src="scTEI_files/figure-html/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make RidgePlot by cell type</span></span>
<span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"cell.type"</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu">RidgePlot</span><span class="op">(</span>object <span class="op">=</span> <span class="va">celegans</span>,</span>
<span>    features <span class="op">=</span> <span class="st">"TEI"</span>,</span>
<span>    group.by <span class="op">=</span> <span class="st">"cell.type"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Picking joint bandwidth of 0.054</span></span></code></pre>
<p><img src="scTEI_files/figure-html/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># subset to specific cell type - ADF + notClassified</span></span>
<span><span class="va">ADF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">celegans</span>,</span>
<span>    cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"ADF"</span>,</span>
<span>        <span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">cell.type</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"notClassified"</span>,</span>
<span>         <span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">cell.type</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p5</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span>object <span class="op">=</span> <span class="va">ADF</span><span class="op">)</span></span>
<span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">ADF</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"embryo.time.bin"</span></span>
<span><span class="va">p6</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span>object <span class="op">=</span> <span class="va">ADF</span><span class="op">)</span></span>
<span><span class="va">p7</span> <span class="op">&lt;-</span> <span class="fu">RidgePlot</span><span class="op">(</span><span class="va">ADF</span>, <span class="st">"TEI"</span><span class="op">)</span></span>
<span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">ADF</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"embryo.time.bin.cell.type"</span></span>
<span><span class="va">p8</span> <span class="op">&lt;-</span> <span class="fu">RidgePlot</span><span class="op">(</span>object <span class="op">=</span> <span class="va">ADF</span>, features <span class="op">=</span> <span class="st">"TEI"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make grid plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">plot_grid</span><span class="op">(</span><span class="va">p5</span>, <span class="va">p6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">plot_grid</span><span class="op">(</span><span class="va">p7</span>, <span class="va">p8</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Picking joint bandwidth of 0.0608</span></span></code></pre>
<pre><code><span><span class="co">## Picking joint bandwidth of 0.0704</span></span></code></pre>
<p><img src="scTEI_files/figure-html/unnamed-chunk-10-2.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use pMatrix as data to cluster</span></span>
<span><span class="va">celegans.TEI</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">celegans</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>,</span>
<span>    meta.data <span class="op">=</span> <span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span>, row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">celegans.TEI</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pMatrixTEI.html">pMatrixTEI</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">celegans.TEI</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span></span>
<span><span class="op">)</span></span>
<span><span class="va">all.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">celegans.TEI</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">celegans.TEI</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span></span>
<span>    <span class="va">celegans.TEI</span>,</span>
<span>    selection.method <span class="op">=</span> <span class="st">"vst"</span>,</span>
<span>    nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>do.scale <span class="op">=</span> <span class="cn">FALSE</span>, do.center <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    features <span class="op">=</span> <span class="va">all.genes</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>dims<span class="op">=</span><span class="fl">50</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Not all features provided are in this Assay object, removing</span></span>
<span><span class="co">## the following feature(s): WBGene00018440, WBGene00013501, WBGene00017610,</span></span>
<span><span class="co">## WBGene00219374, WBGene00008328, WBGene00008881, WBGene00020009, WBGene00010424,</span></span>
<span><span class="co">## WBGene00017211, WBGene00015276, WBGene00219372, WBGene00044439, WBGene00015588</span></span></code></pre>
<pre><code><span><span class="co">## Warning in PrepDR(object = object, features = features, verbose = verbose):</span></span>
<span><span class="co">## The following 88 features requested have zero variance (running reduction</span></span>
<span><span class="co">## without them): WBGene00019055, WBGene00010102, WBGene00010647, WBGene00194713,</span></span>
<span><span class="co">## WBGene00194680, WBGene00235164, WBGene00194689, WBGene00206507, WBGene00045272,</span></span>
<span><span class="co">## WBGene00235158, WBGene00002109, WBGene00021641, WBGene00018204, WBGene00010098,</span></span>
<span><span class="co">## WBGene00001237, WBGene00219421, WBGene00022674, WBGene00011545, WBGene00012760,</span></span>
<span><span class="co">## WBGene00011707, WBGene00017985, WBGene00044562, WBGene00017411, WBGene00206383,</span></span>
<span><span class="co">## WBGene00077522, WBGene00018565, WBGene00021386, WBGene00086553, WBGene00019049,</span></span>
<span><span class="co">## WBGene00195086, WBGene00194832, WBGene00012231, WBGene00206417, WBGene00019251,</span></span>
<span><span class="co">## WBGene00044479, WBGene00010995, WBGene00206379, WBGene00022111, WBGene00019174,</span></span>
<span><span class="co">## WBGene00021804, WBGene00006399, WBGene00018242, WBGene00077493, WBGene00269431,</span></span>
<span><span class="co">## WBGene00011387, WBGene00007642, WBGene00018310, WBGene00022395, WBGene00194952,</span></span>
<span><span class="co">## WBGene00020775, WBGene00015340, WBGene00045208, WBGene00077773, WBGene00018968,</span></span>
<span><span class="co">## WBGene00194684, WBGene00194866, WBGene00009143, WBGene00015339, WBGene00194982,</span></span>
<span><span class="co">## WBGene00194983, WBGene00012301, WBGene00017890, WBGene00219310, WBGene00044502,</span></span>
<span><span class="co">## WBGene00018850, WBGene00010596, WBGene00021966, WBGene00219280, WBGene00012455,</span></span>
<span><span class="co">## WBGene00012214, WBGene00011414, WBGene00015774, WBGene00017584, WBGene00022410,</span></span>
<span><span class="co">## WBGene00018297, WBGene00020040, WBGene00015945, WBGene00012309, WBGene00012254,</span></span>
<span><span class="co">## WBGene00022339, WBGene00022648, WBGene00018234, WBGene00020261, WBGene00043054,</span></span>
<span><span class="co">## WBGene00011934, WBGene00007271, WBGene00006051, WBGene00016724</span></span></code></pre>
<pre><code><span><span class="co">## PC_ 1 </span></span>
<span><span class="co">## Positive:  WBGene00002085, WBGene00009212, WBGene00018352, WBGene00015610, WBGene00022112, WBGene00044066, WBGene00001854, WBGene00012113, WBGene00018606, WBGene00016071 </span></span>
<span><span class="co">##     WBGene00001910, WBGene00008447, WBGene00008066, WBGene00012899, WBGene00019432, WBGene00013553, WBGene00002253, WBGene00014149, WBGene00020537, WBGene00012900 </span></span>
<span><span class="co">##     WBGene00016711, WBGene00012021, WBGene00020205, WBGene00016457, WBGene00012420, WBGene00021299, WBGene00016088, WBGene00019040, WBGene00017713, WBGene00011156 </span></span>
<span><span class="co">## Negative:  WBGene00018480, WBGene00017960, WBGene00018338, WBGene00009626, WBGene00005093, WBGene00011133, WBGene00077680, WBGene00001841, WBGene00013286, WBGene00001202 </span></span>
<span><span class="co">##     WBGene00012708, WBGene00013370, WBGene00007440, WBGene00016682, WBGene00018069, WBGene00020573, WBGene00001614, WBGene00022744, WBGene00021160, WBGene00006664 </span></span>
<span><span class="co">##     WBGene00003100, WBGene00044433, WBGene00006229, WBGene00009593, WBGene00020505, WBGene00009031, WBGene00011296, WBGene00005294, WBGene00016933, WBGene00017032 </span></span>
<span><span class="co">## PC_ 2 </span></span>
<span><span class="co">## Positive:  WBGene00018352, WBGene00009212, WBGene00019040, WBGene00018606, WBGene00001910, WBGene00015610, WBGene00016071, WBGene00001854, WBGene00016711, WBGene00021915 </span></span>
<span><span class="co">##     WBGene00014149, WBGene00017611, WBGene00016463, WBGene00008447, WBGene00019039, WBGene00017326, WBGene00007641, WBGene00007883, WBGene00016297, WBGene00012420 </span></span>
<span><span class="co">##     WBGene00011763, WBGene00044066, WBGene00045386, WBGene00016114, WBGene00016457, WBGene00020772, WBGene00013553, WBGene00022112, WBGene00019432, WBGene00019482 </span></span>
<span><span class="co">## Negative:  WBGene00002085, WBGene00002086, WBGene00015081, WBGene00019516, WBGene00006853, WBGene00001534, WBGene00006047, WBGene00001773, WBGene00018480, WBGene00003055 </span></span>
<span><span class="co">##     WBGene00009386, WBGene00001841, WBGene00017648, WBGene00009593, WBGene00013370, WBGene00015974, WBGene00020675, WBGene00004149, WBGene00000684, WBGene00017960 </span></span>
<span><span class="co">##     WBGene00009626, WBGene00012516, WBGene00011133, WBGene00018338, WBGene00020505, WBGene00006677, WBGene00006897, WBGene00077680, WBGene00006556, WBGene00003414 </span></span>
<span><span class="co">## PC_ 3 </span></span>
<span><span class="co">## Positive:  WBGene00018352, WBGene00009212, WBGene00019040, WBGene00018606, WBGene00016071, WBGene00015610, WBGene00001854, WBGene00008447, WBGene00019039, WBGene00014149 </span></span>
<span><span class="co">##     WBGene00044066, WBGene00022112, WBGene00020537, WBGene00019432, WBGene00020205, WBGene00013553, WBGene00017611, WBGene00008066, WBGene00002253, WBGene00012021 </span></span>
<span><span class="co">##     WBGene00009861, WBGene00000561, WBGene00077531, WBGene00012899, WBGene00021299, WBGene00019035, WBGene00012900, WBGene00007641, WBGene00017713, WBGene00008612 </span></span>
<span><span class="co">## Negative:  WBGene00045386, WBGene00016114, WBGene00016463, WBGene00007883, WBGene00003890, WBGene00009983, WBGene00045336, WBGene00194921, WBGene00016711, WBGene00011289 </span></span>
<span><span class="co">##     WBGene00011500, WBGene00008484, WBGene00011763, WBGene00021915, WBGene00003642, WBGene00010864, WBGene00269432, WBGene00012753, WBGene00016397, WBGene00003934 </span></span>
<span><span class="co">##     WBGene00045331, WBGene00017068, WBGene00012853, WBGene00004346, WBGene00016254, WBGene00044020, WBGene00044568, WBGene00011869, WBGene00019975, WBGene00010848 </span></span>
<span><span class="co">## PC_ 4 </span></span>
<span><span class="co">## Positive:  WBGene00009212, WBGene00044066, WBGene00001910, WBGene00015610, WBGene00022112, WBGene00002253, WBGene00008447, WBGene00018606, WBGene00001854, WBGene00012899 </span></span>
<span><span class="co">##     WBGene00016071, WBGene00016088, WBGene00012021, WBGene00012113, WBGene00013553, WBGene00008066, WBGene00021299, WBGene00016457, WBGene00020537, WBGene00012900 </span></span>
<span><span class="co">##     WBGene00019432, WBGene00014149, WBGene00011156, WBGene00017713, WBGene00020205, WBGene00019039, WBGene00001960, WBGene00010100, WBGene00007934, WBGene00014156 </span></span>
<span><span class="co">## Negative:  WBGene00018352, WBGene00002085, WBGene00017611, WBGene00021915, WBGene00007641, WBGene00016297, WBGene00020772, WBGene00011763, WBGene00019482, WBGene00015062 </span></span>
<span><span class="co">##     WBGene00007883, WBGene00000561, WBGene00016463, WBGene00044068, WBGene00007286, WBGene00021272, WBGene00007490, WBGene00016397, WBGene00009355, WBGene00007675 </span></span>
<span><span class="co">##     WBGene00016532, WBGene00019975, WBGene00016711, WBGene00016254, WBGene00044419, WBGene00008453, WBGene00009983, WBGene00017184, WBGene00011500, WBGene00013899 </span></span>
<span><span class="co">## PC_ 5 </span></span>
<span><span class="co">## Positive:  WBGene00016114, WBGene00016463, WBGene00045336, WBGene00016711, WBGene00007883, WBGene00012753, WBGene00011763, WBGene00044568, WBGene00021915, WBGene00016254 </span></span>
<span><span class="co">##     WBGene00009983, WBGene00044020, WBGene00008484, WBGene00016397, WBGene00003934, WBGene00019975, WBGene00004346, WBGene00164973, WBGene00019177, WBGene00020582 </span></span>
<span><span class="co">##     WBGene00021268, WBGene00016297, WBGene00011869, WBGene00003642, WBGene00011500, WBGene00017076, WBGene00002105, WBGene00015354, WBGene00003745, WBGene00002089 </span></span>
<span><span class="co">## Negative:  WBGene00045386, WBGene00003890, WBGene00018352, WBGene00194921, WBGene00010848, WBGene00045331, WBGene00005246, WBGene00019040, WBGene00016071, WBGene00013899 </span></span>
<span><span class="co">##     WBGene00001854, WBGene00015610, WBGene00019150, WBGene00018606, WBGene00009212, WBGene00019416, WBGene00019039, WBGene00008447, WBGene00010693, WBGene00020537 </span></span>
<span><span class="co">##     WBGene00022112, WBGene00006778, WBGene00020205, WBGene00044066, WBGene00019432, WBGene00014149, WBGene00006985, WBGene00045247, WBGene00017326, WBGene00008066</span></span></code></pre>
<pre><code><span><span class="co">## 09:43:00 UMAP embedding parameters a = 0.9922 b = 1.112</span></span></code></pre>
<pre><code><span><span class="co">## 09:43:00 Read 6188 rows and found 20 numeric columns</span></span></code></pre>
<pre><code><span><span class="co">## 09:43:00 Using Annoy for neighbor search, n_neighbors = 30</span></span></code></pre>
<pre><code><span><span class="co">## 09:43:00 Building Annoy index with metric = cosine, n_trees = 50</span></span></code></pre>
<pre><code><span><span class="co">## 0%   10   20   30   40   50   60   70   80   90   100%</span></span></code></pre>
<pre><code><span><span class="co">## [----|----|----|----|----|----|----|----|----|----|</span></span></code></pre>
<pre><code><span><span class="co">## **************************************************|</span></span>
<span><span class="co">## 09:43:00 Writing NN index file to temp file /var/folders/1_/x30smsdj6mlgnvfrgvx8s4nh0000gp/T//RtmpRgTO9I/file6c0e4d71d9ab</span></span>
<span><span class="co">## 09:43:00 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">## 09:43:02 Annoy recall = 100%</span></span>
<span><span class="co">## 09:43:03 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">## 09:43:04 Initializing from normalized Laplacian + noise (using irlba)</span></span>
<span><span class="co">## 09:43:05 Commencing optimization for 500 epochs, with 247274 positive edges</span></span>
<span><span class="co">## 09:43:15 Optimization finished</span></span></code></pre>
<pre><code><span><span class="co">## Use multiple threads to calculate pMatrix</span></span>
<span></span>
<span><span class="va">celegans.TEI</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pMatrixTEI.html">pMatrixTEI</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">celegans.TEI</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span>,</span>
<span>    split <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">celegans.TEI</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"embryo.time.bin"</span></span>
<span><span class="va">p9</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">celegans.TEI</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p9</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make FeaturePlot</span></span>
<span><span class="va">p10</span> <span class="op">&lt;-</span> <span class="fu">FeaturePlot</span><span class="op">(</span></span>
<span>    object <span class="op">=</span> <span class="va">celegans.TEI</span>,</span>
<span>    features <span class="op">=</span> <span class="st">"TEI"</span>,</span>
<span>    min.cutoff<span class="op">=</span><span class="st">'q05'</span>,</span>
<span>    max.cutoff<span class="op">=</span><span class="st">'q95'</span>,</span>
<span>    cols <span class="op">=</span> <span class="fu">viridis</span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p10</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"cell.type"</span></span>
<span><span class="va">p11</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">celegans</span><span class="op">)</span></span>
<span><span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">celegans.TEI</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"cell.type"</span></span>
<span><span class="va">p12</span> <span class="op">&lt;-</span> <span class="fu">DimPlot</span><span class="op">(</span><span class="va">celegans.TEI</span><span class="op">)</span></span>
<span><span class="co"># make grid plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">plot_grid</span><span class="op">(</span><span class="va">p2</span>, <span class="va">p11</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">plot_grid</span><span class="op">(</span><span class="va">p10</span>, <span class="va">p12</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-14-2.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get TEI per strata</span></span>
<span><span class="va">pS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pStrataTEI.html">pStrataTEI</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">celegans</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Use multiple threads to calculate pStrata</span></span>
<span></span>
<span><span class="va">pS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pStrataTEI.html">pStrataTEI</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">celegans</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>,</span>
<span>    Phylostratum <span class="op">=</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">celegans_ps</span><span class="op">$</span><span class="va">Phylostratum</span><span class="op">)</span>,</span>
<span>        <span class="va">celegans_ps</span><span class="op">$</span><span class="va">GeneID</span><span class="op">)</span>,</span>
<span>    split <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get permutations</span></span>
<span><span class="va">bM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootTEI.html">bootTEI</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">celegans</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span>,</span>
<span>    permutations <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Use multiple threads to get permutations</span></span>
<span></span>
<span><span class="va">bM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootTEI.html">bootTEI</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">celegans</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span>,</span>
<span>    permutations <span class="op">=</span> <span class="fl">100</span>,</span>
<span>    split <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    threads <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get mean expression matrix</span></span>
<span><span class="va">meanMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/REMatrix.html">REMatrix</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">celegans</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get mean expression matrix with groups</span></span>
<span><span class="va">cell_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">cell.type</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">cell.type</span><span class="op">==</span><span class="va">x</span><span class="op">)</span><span class="op">}</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">cell.type</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">meanMatrix_by_cell.type</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/REMatrix.html">REMatrix</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">celegans</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">scale.data</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span>,</span>
<span>    groups <span class="op">=</span> <span class="va">cell_groups</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">meanMatrix_by_cell.type</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, cluster_columns <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    col <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># computing relative expression profile over cell types</span></span>
<span><span class="va">cell_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">cell.type</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">cell.type</span><span class="op">==</span><span class="va">x</span><span class="op">)</span><span class="op">}</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">celegans</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">cell.type</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">reMatrix_by_cell.type</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/REMatrix.html">REMatrix</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="va">celegans</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">scale.data</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span>,</span>
<span>    groups <span class="op">=</span> <span class="va">cell_groups</span>,</span>
<span>    by <span class="op">=</span> <span class="st">"row"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">reMatrix_by_cell.type</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="monocle3---example">Monocle3 - example<a class="anchor" aria-label="anchor" href="#monocle3---example"></a>
</h2>
<p>For <code>Monocle3</code>, we again use scRNA data obtained from
<span class="citation">(Packer et al. 2019)</span> for the model
organism <em>Caenorhabditis elegans</em> <span class="citation">(Packer
et al. 2019)</span> and phylogenetic data obtained from <span class="citation">(Sun, Rödelsperger, and Sommer 2021)</span>
(Sun2021_Supplemental_Table_S6).</p>
<div class="section level3">
<h3 id="example-6k-c--elegans-embryos-from-packer-and-zhu-et-al-2019-1">Example: 6k C. elegans embryos from Packer and Zhu et al (2019)<a class="anchor" aria-label="anchor" href="#example-6k-c--elegans-embryos-from-packer-and-zhu-et-al-2019-1"></a>
</h3>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load Packer and Zhu et al (2019) data set</span></span>
<span><span class="va">expression_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"http://staff.washington.edu/hpliner/data/"</span>,</span>
<span>    <span class="st">"packer_embryo_expression.rds"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">cell_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"http://staff.washington.edu/hpliner/data/"</span>,</span>
<span>    <span class="st">"packer_embryo_colData.rds"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">gene_annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"http://staff.washington.edu/hpliner/data/"</span>,</span>
<span>    <span class="st">"packer_embryo_rowData.rds"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">new_cell_data_set</span><span class="op">(</span></span>
<span>    expression_data <span class="op">=</span> <span class="va">expression_matrix</span>,</span>
<span>    cell_metadata <span class="op">=</span> <span class="va">cell_metadata</span>,</span>
<span>    gene_metadata <span class="op">=</span> <span class="va">gene_annotation</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># preprocess scRNA</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">preprocess_cds</span><span class="op">(</span><span class="va">cds</span>, num_dim <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">align_cds</span><span class="op">(</span><span class="va">cds</span>, alignment_group <span class="op">=</span> <span class="st">"batch"</span>,</span>
<span>    residual_model_formula_str <span class="op">=</span> <span class="st">"~ bg.300.loading +</span></span>
<span><span class="st">        bg.400.loading + bg.500.1.loading + bg.500.2.loading +</span></span>
<span><span class="st">        bg.r17.loading + bg.b01.loading + bg.b02.loading"</span><span class="op">)</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">reduce_dimension</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">cluster_cells</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">learn_graph</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span></span></code></pre></div>
<pre><code>## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |======================================================================| 100%</code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">plot_cells</span><span class="op">(</span><span class="va">cds</span>,</span>
<span>    label_groups_by_cluster<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>    color_cells_by <span class="op">=</span> <span class="st">"embryo.time.bin"</span>,</span>
<span>    group_label_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    label_cell_groups <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    label_leaves <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    label_branch_points <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    graph_label_size<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p1</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-22-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># order cells - select youngest time point according to embryo.time</span></span>
<span><span class="co"># - center of the plot</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">order_cells</span><span class="op">(</span><span class="va">cds</span>,</span>
<span>    root_cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">cds</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span></span>
<span>    <span class="fu">colData</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span><span class="op">$</span><span class="va">embryo.time</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span><span class="op">$</span><span class="va">embryo.time</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span><span class="op">[</span><span class="st">"pseudotime"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">pseudotime</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot by pseudotime</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">plot_cells</span><span class="op">(</span><span class="va">cds</span>,</span>
<span>    label_groups_by_cluster<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>    color_cells_by <span class="op">=</span> <span class="st">"pseudotime"</span>,</span>
<span>    group_label_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    label_cell_groups <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    label_leaves <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    label_branch_points <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    graph_label_size<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p2</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-23-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load Caenorhabditis elegans gene age estimation</span></span>
<span><span class="va">celegans_ps</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>    <span class="st">"Sun2021_Orthomap.tsv"</span>, package <span class="op">=</span> <span class="st">"scTEI"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">20040</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">2</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">## <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">## <span style="color: #BB0000;">chr</span> (1): GeneID</span></span>
<span><span class="co">## <span style="color: #00BB00;">dbl</span> (1): Phylostratum</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">celegans_ps</span><span class="op">$</span><span class="va">Phylostratum</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##    0    1    2    3    4    5    6    7    8    9   10   11   12   13 </span></span>
<span><span class="co">## 1290 5434 4666  603 1039  808  274  884  590  511  384 1113 1277 1167</span></span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># define Phylostratum</span></span>
<span><span class="va">ps_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">celegans_ps</span><span class="op">$</span><span class="va">Phylostratum</span><span class="op">)</span>,</span>
<span>    <span class="va">celegans_ps</span><span class="op">$</span><span class="va">GeneID</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add TEI values</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span><span class="op">[</span><span class="st">"TEI"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TEI.html">TEI</a></span><span class="op">(</span></span>
<span>    ExpressionSet <span class="op">=</span> <span class="fu">counts</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span>,</span>
<span>    Phylostratum <span class="op">=</span> <span class="va">ps_vec</span></span>
<span>        <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Use multiple threads to calculate TEI on sparseMatrix</span></span>
<span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span><span class="op">[</span><span class="st">"TEI"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TEI.html">TEI</a></span><span class="op">(</span>ExpressionSet <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">cds</span><span class="op">)</span>,</span>
<span>    Phylostratum <span class="op">=</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">celegans_ps</span><span class="op">$</span><span class="va">Phylostratum</span><span class="op">)</span>,</span>
<span>        <span class="va">celegans_ps</span><span class="op">$</span><span class="va">GeneID</span><span class="op">)</span>,</span>
<span>    split <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make FeaturePlot</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">plot_cells</span><span class="op">(</span><span class="va">cds</span>,</span>
<span>    label_groups_by_cluster<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>    color_cells_by <span class="op">=</span> <span class="st">"TEI"</span>,</span>
<span>    group_label_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    label_cell_groups <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    label_leaves <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    label_branch_points <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    graph_label_size<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>    cell_size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p3</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make Boxplot</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">embryo.time.bin</span>, y<span class="op">=</span><span class="va">TEI</span>, fill<span class="op">=</span><span class="va">embryo.time.bin</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_violin</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_boxplot</span><span class="op">(</span>width<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p4</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-27-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make scatter plot - TEI vs pseudotime</span></span>
<span><span class="va">p5</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">pseudotime</span>, y<span class="op">=</span><span class="va">TEI</span>, col<span class="op">=</span><span class="va">pseudotime</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p5</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-28-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p6</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">pseudotime</span>, y<span class="op">=</span><span class="va">TEI</span>, col<span class="op">=</span><span class="va">embryo.time.bin</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p6</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-28-2.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make grid plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">plot_grid</span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span>, <span class="va">p4</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="scTEI_files/figure-html/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-cantalapiedra2021eggnog" class="csl-entry">
Cantalapiedra, Carlos P, Ana Hernández-Plaza, Ivica Letunic, Peer Bork,
and Jaime Huerta-Cepas. 2021. <span>“eggNOG-Mapper V2: Functional
Annotation, Orthology Assignments, and Domain Prediction at the
Metagenomic Scale.”</span> <em>Molecular Biology and Evolution</em> 38
(12): 5825–29.
</div>
<div id="ref-cao2019single" class="csl-entry">
Cao, Junyue, Malte Spielmann, Xiaojie Qiu, Xingfan Huang, Daniel M
Ibrahim, Andrew J Hill, Fan Zhang, et al. 2019. <span>“The Single-Cell
Transcriptional Landscape of Mammalian Organogenesis.”</span>
<em>Nature</em> 566 (7745): 496–502.
</div>
<div id="ref-cazet2022new" class="csl-entry">
Cazet, Jack, Stefan Siebert, Hannah Morris Little, Philip Bertemes, Abby
S Primack, Peter Ladurner, Matthias Achrainer, et al. 2022. <span>“New
Hydra Genomes Reveal Conserved Principles of Hydrozoan Transcriptional
Regulation.”</span> <em>bioRxiv</em>.
</div>
<div id="ref-domazet2007phylostratigraphy" class="csl-entry">
Domazet-Lošo, Tomislav, Josip Brajković, and Diethard Tautz. 2007.
<span>“A Phylostratigraphy Approach to Uncover the Genomic History of
Major Adaptations in Metazoan Lineages.”</span> <em>Trends in
Genetics</em> 23 (11): 533–39.
</div>
<div id="ref-domazet2010phylogenetically" class="csl-entry">
Domazet-Lošo, Tomislav, and Diethard Tautz. 2010. <span>“A
Phylogenetically Based Transcriptome Age Index Mirrors Ontogenetic
Divergence Patterns.”</span> <em>Nature</em> 468 (7325): 815–18.
</div>
<div id="ref-drost2017cross" class="csl-entry">
Drost, Hajk-Georg, Philipp Janitza, Ivo Grosse, and Marcel Quint. 2017.
<span>“Cross-Kingdom Comparison of the Developmental Hourglass.”</span>
<em>Current Opinion in Genetics &amp; Development</em> 45: 69–75.
</div>
<div id="ref-emms2019orthofinder" class="csl-entry">
Emms, David M, and Steven Kelly. 2019. <span>“OrthoFinder: Phylogenetic
Orthology Inference for Comparative Genomics.”</span> <em>Genome
Biology</em> 20 (1): 1–14.
</div>
<div id="ref-gossmann2016transcriptomes" class="csl-entry">
Gossmann, Toni I, Dounia Saleh, Marc W Schmid, Michael A Spence, and
Karl J Schmid. 2016. <span>“Transcriptomes of Plant Gametophytes Have a
Higher Proportion of Rapidly Evolving and Young Genes Than
Sporophytes.”</span> <em>Molecular Biology and Evolution</em> 33 (7):
1669–78.
</div>
<div id="ref-huerta2019eggnog" class="csl-entry">
Huerta-Cepas, Jaime, Damian Szklarczyk, Davide Heller, Ana
Hernández-Plaza, Sofia K Forslund, Helen Cook, Daniel R Mende, et al.
2019. <span>“eggNOG 5.0: A Hierarchical, Functionally and
Phylogenetically Annotated Orthology Resource Based on 5090 Organisms
and 2502 Viruses.”</span> <em>Nucleic Acids Research</em> 47 (D1):
D309–14.
</div>
<div id="ref-julca2021comparative" class="csl-entry">
Julca, Irene, Camilla Ferrari, Marı́a Flores-Tornero, Sebastian Proost,
Ann-Cathrin Lindner, Dieter Hackenberg, Lenka Steinbachová, et al. 2021.
<span>“Comparative Transcriptomic Analysis Reveals Conserved Programmes
Underpinning Organogenesis and Reproduction in Land Plants.”</span>
<em>Nature Plants</em> 7 (8): 1143–59.
</div>
<div id="ref-lechner2011proteinortho" class="csl-entry">
Lechner, Marcus, Sven Findeiß, Lydia Steiner, Manja Marz, Peter F
Stadler, and Sonja J Prohaska. 2011. <span>“Proteinortho: Detection of
(Co-) Orthologs in Large-Scale Analysis.”</span> <em>BMC
Bioinformatics</em> 12 (1): 1–9.
</div>
<div id="ref-linard2021ten" class="csl-entry">
Linard, Benjamin, Ingo Ebersberger, Shawn E McGlynn, Natasha Glover,
Tomohiro Mochizuki, Mateus Patricio, Odile Lecompte, et al. 2021.
<span>“Ten Years of Collaborative Progress in the Quest for
Orthologs.”</span> <em>Molecular Biology and Evolution</em> 38 (8):
3033–45.
</div>
<div id="ref-liu2020inter" class="csl-entry">
Liu, Jialin, Michael Frochaux, Vincent Gardeux, Bart Deplancke, and Marc
Robinson-Rechavi. 2020. <span>“Inter-Embryo Gene Expression Variability
Recapitulates the Hourglass Pattern of Evo-Devo.”</span> <em>BMC
Biology</em> 18 (1): 1–12.
</div>
<div id="ref-packer2019lineage" class="csl-entry">
Packer, Jonathan S, Qin Zhu, Chau Huynh, Priya Sivaramakrishnan, Elicia
Preston, Hannah Dueck, Derek Stefanik, et al. 2019. <span>“A
Lineage-Resolved Molecular Atlas of c. Elegans Embryogenesis at
Single-Cell Resolution.”</span> <em>Science</em> 365 (6459): eaax1971.
</div>
<div id="ref-proost2009plaza" class="csl-entry">
Proost, Sebastian, Michiel Van Bel, Lieven Sterck, Kenny Billiau, Thomas
Van Parys, Yves Van de Peer, and Klaas Vandepoele. 2009. <span>“PLAZA: A
Comparative Genomics Resource to Study Gene and Genome Evolution in
Plants.”</span> <em>The Plant Cell</em> 21 (12): 3718–31.
</div>
<div id="ref-quint2012transcriptomic" class="csl-entry">
Quint, Marcel, Hajk-Georg Drost, Alexander Gabel, Kristian Karsten
Ullrich, Markus Bönn, and Ivo Grosse. 2012. <span>“A Transcriptomic
Hourglass in Plant Embryogenesis.”</span> <em>Nature</em> 490 (7418):
98–101.
</div>
<div id="ref-satija2015spatial" class="csl-entry">
Satija, Rahul, Jeffrey A Farrell, David Gennert, Alexander F Schier, and
Aviv Regev. 2015. <span>“Spatial Reconstruction of Single-Cell Gene
Expression Data.”</span> <em>Nature Biotechnology</em> 33 (5): 495–502.
</div>
<div id="ref-schneider2007oma" class="csl-entry">
Schneider, Adrian, Christophe Dessimoz, and Gaston H Gonnet. 2007.
<span>“OMA Browser—Exploring Orthologous Relations Across 352 Complete
Genomes.”</span> <em>Bioinformatics</em> 23 (16): 2180–82.
</div>
<div id="ref-sun2021single" class="csl-entry">
Sun, Shuai, Christian Rödelsperger, and Ralf J Sommer. 2021.
<span>“Single Worm Transcriptomics Identifies a Developmental Core
Network of Oscillating Genes with Deep Conservation Across
Nematodes.”</span> <em>Genome Research</em> 31 (9): 1590–1601.
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2022-01-07 r81454)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Big Sur/Monterey 10.16</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    grid      stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] monocle3_1.2.9                   SingleCellExperiment_1.19.1     </span></span>
<span><span class="co">##  [3] SummarizedExperiment_1.27.3      GenomicRanges_1.49.1            </span></span>
<span><span class="co">##  [5] GenomeInfoDb_1.33.12             IRanges_2.31.2                  </span></span>
<span><span class="co">##  [7] S4Vectors_0.35.4                 MatrixGenerics_1.9.1            </span></span>
<span><span class="co">##  [9] matrixStats_0.62.0               Biobase_2.57.1                  </span></span>
<span><span class="co">## [11] BiocGenerics_0.43.4              celegans.embryo.SeuratData_0.1.0</span></span>
<span><span class="co">## [13] SeuratData_0.2.1                 sp_1.5-0                        </span></span>
<span><span class="co">## [15] SeuratObject_4.1.2               Seurat_4.2.0                    </span></span>
<span><span class="co">## [17] ComplexHeatmap_2.13.3            ggplot2_3.3.6                   </span></span>
<span><span class="co">## [19] cowplot_1.1.1                    viridis_0.6.2                   </span></span>
<span><span class="co">## [21] viridisLite_0.4.1                readr_2.1.3                     </span></span>
<span><span class="co">## [23] plyr_1.8.7                       dplyr_1.0.10                    </span></span>
<span><span class="co">## [25] scTEI_0.1.0                      BiocStyle_2.25.0                </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] utf8_1.2.2                reticulate_1.26          </span></span>
<span><span class="co">##   [3] tidyselect_1.2.0          lme4_1.1-30              </span></span>
<span><span class="co">##   [5] htmlwidgets_1.5.4         BiocParallel_1.31.13     </span></span>
<span><span class="co">##   [7] Rtsne_0.16                ScaledMatrix_1.5.1       </span></span>
<span><span class="co">##   [9] munsell_0.5.0             codetools_0.2-18         </span></span>
<span><span class="co">##  [11] ragg_1.2.3                ica_1.0-3                </span></span>
<span><span class="co">##  [13] future_1.28.0             miniUI_0.1.1.1           </span></span>
<span><span class="co">##  [15] batchelor_1.13.3          withr_2.5.0              </span></span>
<span><span class="co">##  [17] spatstat.random_2.2-0     colorspace_2.0-3         </span></span>
<span><span class="co">##  [19] progressr_0.11.0          highr_0.9                </span></span>
<span><span class="co">##  [21] knitr_1.40                rstudioapi_0.14          </span></span>
<span><span class="co">##  [23] ROCR_1.0-11               tensor_1.5               </span></span>
<span><span class="co">##  [25] listenv_0.8.0             labeling_0.4.2           </span></span>
<span><span class="co">##  [27] GenomeInfoDbData_1.2.9    polyclip_1.10-0          </span></span>
<span><span class="co">##  [29] bit64_4.0.5               farver_2.1.1             </span></span>
<span><span class="co">##  [31] rprojroot_2.0.3           parallelly_1.32.1        </span></span>
<span><span class="co">##  [33] vctrs_0.4.2               generics_0.1.3           </span></span>
<span><span class="co">##  [35] xfun_0.34                 R6_2.5.1                 </span></span>
<span><span class="co">##  [37] doParallel_1.0.17         ggbeeswarm_0.6.0         </span></span>
<span><span class="co">##  [39] clue_0.3-62               rsvd_1.0.5               </span></span>
<span><span class="co">##  [41] bitops_1.0-7              spatstat.utils_3.0-1     </span></span>
<span><span class="co">##  [43] cachem_1.0.6              DelayedArray_0.23.2      </span></span>
<span><span class="co">##  [45] assertthat_0.2.1          vroom_1.6.0              </span></span>
<span><span class="co">##  [47] promises_1.2.0.1          scales_1.2.1             </span></span>
<span><span class="co">##  [49] beeswarm_0.4.0            rgeos_0.5-9              </span></span>
<span><span class="co">##  [51] gtable_0.3.1              beachmat_2.13.4          </span></span>
<span><span class="co">##  [53] Cairo_1.6-0               globals_0.16.1           </span></span>
<span><span class="co">##  [55] goftest_1.2-3             rlang_1.0.6              </span></span>
<span><span class="co">##  [57] systemfonts_1.0.4         GlobalOptions_0.1.2      </span></span>
<span><span class="co">##  [59] splines_4.2.0             lazyeval_0.2.2           </span></span>
<span><span class="co">##  [61] spatstat.geom_2.4-0       BiocManager_1.30.18      </span></span>
<span><span class="co">##  [63] yaml_2.3.6                reshape2_1.4.4           </span></span>
<span><span class="co">##  [65] abind_1.4-5               httpuv_1.6.6             </span></span>
<span><span class="co">##  [67] tools_4.2.0               bookdown_0.29            </span></span>
<span><span class="co">##  [69] ellipsis_0.3.2            spatstat.core_2.4-4      </span></span>
<span><span class="co">##  [71] jquerylib_0.1.4           RColorBrewer_1.1-3       </span></span>
<span><span class="co">##  [73] proxy_0.4-27              ggridges_0.5.4           </span></span>
<span><span class="co">##  [75] Rcpp_1.0.9                sparseMatrixStats_1.9.0  </span></span>
<span><span class="co">##  [77] zlibbioc_1.43.0           purrr_0.3.5              </span></span>
<span><span class="co">##  [79] RCurl_1.98-1.9            rpart_4.1.16             </span></span>
<span><span class="co">##  [81] deldir_1.0-6              pbapply_1.5-0            </span></span>
<span><span class="co">##  [83] GetoptLong_1.0.5          zoo_1.8-11               </span></span>
<span><span class="co">##  [85] ggrepel_0.9.1             cluster_2.1.4            </span></span>
<span><span class="co">##  [87] fs_1.5.2                  magrittr_2.0.3           </span></span>
<span><span class="co">##  [89] data.table_1.14.4         scattermore_0.8          </span></span>
<span><span class="co">##  [91] ResidualMatrix_1.7.1      circlize_0.4.15          </span></span>
<span><span class="co">##  [93] lmtest_0.9-40             RANN_2.6.1               </span></span>
<span><span class="co">##  [95] myTAI_0.9.3               fitdistrplus_1.1-8       </span></span>
<span><span class="co">##  [97] hms_1.1.2                 patchwork_1.1.2          </span></span>
<span><span class="co">##  [99] mime_0.12                 evaluate_0.17            </span></span>
<span><span class="co">## [101] xtable_1.8-4              gridExtra_2.3            </span></span>
<span><span class="co">## [103] shape_1.4.6               compiler_4.2.0           </span></span>
<span><span class="co">## [105] tibble_3.1.8              KernSmooth_2.23-20       </span></span>
<span><span class="co">## [107] crayon_1.5.2              minqa_1.2.5              </span></span>
<span><span class="co">## [109] htmltools_0.5.3           mgcv_1.8-40              </span></span>
<span><span class="co">## [111] later_1.3.0               tzdb_0.3.0               </span></span>
<span><span class="co">## [113] tidyr_1.2.1               DBI_1.1.3                </span></span>
<span><span class="co">## [115] MASS_7.3-58.1             rappdirs_0.3.3           </span></span>
<span><span class="co">## [117] boot_1.3-28               leidenbase_0.1.9         </span></span>
<span><span class="co">## [119] Matrix_1.5-1              cli_3.4.1                </span></span>
<span><span class="co">## [121] parallel_4.2.0            igraph_1.3.5             </span></span>
<span><span class="co">## [123] pkgconfig_2.0.3           pkgdown_2.0.6            </span></span>
<span><span class="co">## [125] scuttle_1.7.4             terra_1.6-17             </span></span>
<span><span class="co">## [127] plotly_4.10.0             spatstat.sparse_2.1-1    </span></span>
<span><span class="co">## [129] foreach_1.5.2             vipor_0.4.5              </span></span>
<span><span class="co">## [131] bslib_0.4.0               XVector_0.37.1           </span></span>
<span><span class="co">## [133] stringr_1.4.1             digest_0.6.30            </span></span>
<span><span class="co">## [135] sctransform_0.3.5         RcppAnnoy_0.0.19         </span></span>
<span><span class="co">## [137] spatstat.data_2.2-0       rmarkdown_2.17           </span></span>
<span><span class="co">## [139] leiden_0.4.3              uwot_0.1.14              </span></span>
<span><span class="co">## [141] DelayedMatrixStats_1.19.2 shiny_1.7.2              </span></span>
<span><span class="co">## [143] nloptr_2.0.3              rjson_0.2.21             </span></span>
<span><span class="co">## [145] lifecycle_1.0.3           nlme_3.1-160             </span></span>
<span><span class="co">## [147] jsonlite_1.8.2            BiocNeighbors_1.15.1     </span></span>
<span><span class="co">## [149] limma_3.53.10             desc_1.4.2               </span></span>
<span><span class="co">## [151] fansi_1.0.3               pillar_1.8.1             </span></span>
<span><span class="co">## [153] lattice_0.20-45           ggrastr_1.0.1            </span></span>
<span><span class="co">## [155] fastmap_1.1.0             httr_1.4.4               </span></span>
<span><span class="co">## [157] survival_3.4-0            glue_1.6.2               </span></span>
<span><span class="co">## [159] png_0.1-7                 iterators_1.0.14         </span></span>
<span><span class="co">## [161] bit_4.0.4                 stringi_1.7.8            </span></span>
<span><span class="co">## [163] sass_0.4.2                textshaping_0.3.6        </span></span>
<span><span class="co">## [165] BiocSingular_1.13.1       memoise_2.0.1            </span></span>
<span><span class="co">## [167] irlba_2.3.5.1             future.apply_1.9.1</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kristian K Ullrich.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
