# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @useDynLib scTEI, .registration = TRUE
#' @import Rcpp
#' @import Matrix
#' @title rcpp_boottei_parallel
#' @name rcpp_boottei_parallel
#' @description computes the phylogenetically based
#' transcriptome evolutionary index (TEI) shuffling the strata for permutation
#' statistic
#' @return sparseMatrix
#' @param expression ExpressionSet as sparseMatrix
#' @param ps named Phylostratum
#' @param permutations number of permutations
#' @param ncores number of cores
#' @examples
#' ## load example PhyloExpressionSetExample
#'
#' data("PhyloExpressionSetExample", package="myTAI")
#'
#' ## convert into sparseMatrix - rownames GeneID
#'
#' spmat <- as(data.matrix(PhyloExpressionSetExample[,-c(1,2)]),
#'     "sparseMatrix")
#' rownames(spmat) <- PhyloExpressionSetExample$GeneID
#'
#' ## create named Phylostratum vector
#'
#' ps <- setNames(PhyloExpressionSetExample$Phylostratum,
#'     PhyloExpressionSetExample$GeneID)
#'
#' ## get permutations
#' rcpp_boottei_parallel(spmat, ps, 100, 1)
#' @export rcpp_boottei_parallel
#' @author Kristian K Ullrich
NULL

rcpp_boottei_parallel <- function(expression, ps, permutations, ncores = 1L) {
    .Call(`_scTEI_rcpp_boottei_parallel`, expression, ps, permutations, ncores)
}

#' @useDynLib scTEI, .registration = TRUE
#' @import Rcpp
#' @import Matrix
#' @title rcpp_meanMatrix_parallel
#' @name rcpp_meanMatrix_parallel
#' @description computes mean expression
#' @return sparseMatrix
#' @param expression ExpressionSet as sparseMatrix
#' @param ps named Phylostratum
#' @param psgroup ordered unique Phylostratum
#' @param ncores number of cores
#' @examples
#' ## load example PhyloExpressionSetExample
#'
#' data("PhyloExpressionSetExample", package="myTAI")
#'
#' ## convert into sparseMatrix - rownames GeneID
#'
#' spmat <- as(data.matrix(PhyloExpressionSetExample[,-c(1,2)]),
#'     "sparseMatrix")
#' rownames(spmat) <- PhyloExpressionSetExample$GeneID
#'
#' ## create named Phylostratum vector
#'
#' ps <- setNames(PhyloExpressionSetExample$Phylostratum,
#'     PhyloExpressionSetExample$GeneID)
#' psgroup <- sort(unique(ps))
#'
#' ## get meanMatrix
#' rcpp_meanMatrix_parallel(spmat, ps, psgroup)
#' @export rcpp_meanMatrix_parallel
#' @author Kristian K Ullrich
rcpp_meanMatrix_parallel <- function(expression, ps, psgroup, ncores = 1L) {
    .Call(`_scTEI_rcpp_meanMatrix_parallel`, expression, ps, psgroup, ncores)
}

#' @useDynLib scTEI, .registration = TRUE
#' @import Rcpp
#' @import Matrix
#' @title rcpp_pMatrix_parallel
#' @name rcpp_pMatrix_parallel
#' @description computes the partial
#' transcriptome evolutionary index (TEI) values for each single gene
#' @return sparseMatrix
#' @param expression ExpressionSet as sparseMatrix
#' @param ps named Phylostratum
#' @param ncores number of cores
#' @examples
#' ## load example PhyloExpressionSetExample
#'
#' data("PhyloExpressionSetExample", package="myTAI")
#'
#' ## convert into sparseMatrix - rownames GeneID
#'
#' spmat <- as(data.matrix(PhyloExpressionSetExample[,-c(1,2)]),
#'     "sparseMatrix")
#' rownames(spmat) <- PhyloExpressionSetExample$GeneID
#'
#' ## create named Phylostratum vector
#'
#' ps <- setNames(PhyloExpressionSetExample$Phylostratum,
#'     PhyloExpressionSetExample$GeneID)
#'
#' ## get pMatrix
#' rcpp_pMatrix_parallel(spmat, ps)
#' @export rcpp_pMatrix_parallel
#' @author Kristian K Ullrich
rcpp_pMatrix_parallel <- function(expression, ps, ncores = 1L) {
    .Call(`_scTEI_rcpp_pMatrix_parallel`, expression, ps, ncores)
}

#' @useDynLib scTEI, .registration = TRUE
#' @import Rcpp
#' @import Matrix
#' @title rcpp_pStrata_parallel
#' @name rcpp_pStrata_parallel
#' @description computes the partial
#' transcriptome evolutionary index (TEI) values combined into strata
#' @return sparseMatrix
#' @param expression ExpressionSet as sparseMatrix
#' @param ps named Phylostratum
#' @param psgroup ordered unique Phylostratum
#' @param ncores number of cores
#' @examples
#' ## load example PhyloExpressionSetExample
#'
#' data("PhyloExpressionSetExample", package="myTAI")
#'
#' ## convert into sparseMatrix - rownames GeneID
#'
#' spmat <- as(data.matrix(PhyloExpressionSetExample[,-c(1,2)]),
#'     "sparseMatrix")
#' rownames(spmat) <- PhyloExpressionSetExample$GeneID
#'
#' ## create named Phylostratum vector
#'
#' ps <- setNames(PhyloExpressionSetExample$Phylostratum,
#'     PhyloExpressionSetExample$GeneID)
#' psgroup <- sort(unique(ps))
#'
#' ## get pStrata
#' rcpp_pStrata_parallel(spmat, ps, psgroup)
#' @export rcpp_pStrata_parallel
#' @author Kristian K Ullrich
rcpp_pStrata_parallel <- function(expression, ps, psgroup, ncores = 1L) {
    .Call(`_scTEI_rcpp_pStrata_parallel`, expression, ps, psgroup, ncores)
}

#' @useDynLib scTEI, .registration = TRUE
#' @import Rcpp
#' @import Matrix
#' @title rcpp_tei_parallel
#' @name rcpp_tei_parallel
#' @description computes the phylogenetically based
#' transcriptome evolutionary index (TEI)
#' @return list
#' @param expression ExpressionSet as sparseMatrix
#' @param ps named Phylostratum
#' @param ncores number of cores
#' @examples
#' ## load example PhyloExpressionSetExample
#'
#' data("PhyloExpressionSetExample", package="myTAI")
#'
#' ## convert into sparseMatrix - rownames GeneID
#'
#' spmat <- as(data.matrix(PhyloExpressionSetExample[,-c(1,2)]),
#'     "sparseMatrix")
#' rownames(spmat) <- PhyloExpressionSetExample$GeneID
#'
#' ## create named Phylostratum vector
#'
#' ps <- setNames(PhyloExpressionSetExample$Phylostratum,
#'     PhyloExpressionSetExample$GeneID)
#'
#' ## get TEI
#' rcpp_tei_parallel(spmat, ps)
#' @export rcpp_tei_parallel
#' @author Kristian K Ullrich
rcpp_tei_parallel <- function(expression, ps, ncores = 1L) {
    .Call(`_scTEI_rcpp_tei_parallel`, expression, ps, ncores)
}

